src/mscorlib/src/System/Span.cs

```diff
@@ -420,5 +439,49 @@ internal static unsafe void CopyTo<T>(ref T destination, ref T source, int eleme
                 }
             }
         }
+
+        internal static unsafe void Fill<T>(ref T destination, T value, int elementsCount)
+        {
+            if (!JitHelpers.ContainsReferences<T>())
+            {
+                int size = Unsafe.SizeOf<T>();
+
+                int copiedCount; // Fill up to copiedCount in first loop
+                if (size <= sizeof(int))
+                    copiedCount = elementsCount <= 512 / 4 * 3 ? elementsCount : 512 / 4 * 2;
+                else
+                    copiedCount = elementsCount <= (512 * 3 + size - 1) / size ? elementsCount : (512 * 2 + size - 1) / size;
+
+                for (int i = copiedCount - 1; i >= 0; i--)
+                    Unsafe.Add(ref destination, i) = value;
+
+                elementsCount -= copiedCount;
+                if (elementsCount <= 0)
+                    return;
+
+                fixed (byte* pSource = &Unsafe.As<T, byte>(ref destination))
+                {
+                    while (elementsCount > 0)
+                    {
+                        int copyLen = Math.Min(copiedCount, elementsCount);
+                        fixed (byte* pDestination = &Unsafe.As<T, byte>(ref Unsafe.Add(ref destination, copiedCount)))
+                        {
+#if BIT64
+                            Buffer.Memmove(pDestination, pSource, (ulong)copyLen * (ulong)size);
```

Comment:
This will work ok for smaller sizes where everything fits into the cache. I would expect that this is going to be slower for larger sizes because of it will be faulting twice as many cache lines than simple loop.